### DISCUSSION TITLE
Custom characteristics

### DISCUSSION THREAD

**Discussion by danricho** [HomeSpan/HomeSpan#138] [Q&A] [answered]
Has anyone created a custom characteristic which can be adjusted in (for example) the Eve app. Basically a non-Apple characteristic but is still similar to the built-ins.

I know there will need to be modifications to the library to do this.

The example I'd like to do is to make the "favourite" brightness in Tutorial 15 a characteristic very similar to the Brightness characteristic that isn't shown in the Home app but can be adjusted in Eve and I would also make it stored in NVS (I believe just adding the true argument at initialisation).

I've started diving into it myself but I'm hoping someone has done this already and can provide something for me to work from!

Cheers!

**Comment by danricho** [comment]
I am thinking that adding a new entry in `characteristics.h` like:  
`HAPCHAR( BrightnessSaved, 8, PR+PW+EV, INT, false );`

this is a macro shortcut as follows:
`#define HAPCHAR(hapName,type,perms,format,staticRange)  HapChar hapName {#type,#hapName,(PERMS)(perms),format,staticRange}`

I'm still looking into whether a random `type` can be used and what the `perms` define.. that might be all that I need to do to stop it being shown as a second brightness in iOS Home.

**Comment by danricho** [comment]
Small update.

I've modified the `favourite` in the Tutorial 15 to be a characteristic and I'm able to restore its value upon a power cycle by adding `true` argument.

I'm yet to know how to get the characteristic to show in Eve.

The changes I made to do this:

```cpp
// in 'Characteristics.h'

// added:
HAPCHAR( BrightnessSaved, 9, PR+PW+EV, INT, false );
// to line 74 
 // I adjusted the 'type' to 9 instead of 8. I'm unsure if this is important.
```
```cpp
// in 'Span.h'

// added:
OPT(BrightnessSaved);
// to line 223 

// added
CREATE_CHAR(int,BrightnessSaved,0,0,100);
// to line 396 in 'Span.h'
```

I have however found that this may have introduced some instability where the accessory become unavailable sometimes. I'm still looking into why. It may be because of the UUID that the characteristic is using.

**Comment by HomeSpan** [comment, marked as answer]
Unfortunately, HomeKit Accessories can only _implement_ Services and Characteristics - they can't _define_ new ones.  This is because Services and Characteristics are more formally defined (created) in the application that controls the Accessories.  The Home App has its list of defined Services and Characteristics (as documented in HAP-R2).  Eve uses all the standard HomeKit definitions and apparently adds some new Services and Characteristics, but you need to know the full UUID codes in order to use them.  The short codes (2-3 hex digits) used by HomeKit are reserved just for Apple's Services and Characteristics.   Eve does not seem to publish a list of their UUID codes, though I think others have done some reverse-engineering and published some Eve codes on various blogs.

If you just create a new Characteristic with the short version of a UUID, the Home App will likely complain (or simply say no-response) since it won't recognize the code.  Using the full UUID ensures the Home App will ignore the Characteristic.  See page 60 of the HAP-R2 document for more details on UUIDs.

Since HomeSpan stores Characteristic types as (char *) strings you should be able to implement a new Characteristic using the full UUID, though some modifications to the code may be necessary (I'll have to check).  The key is finding an Eve Characteristic that provides the function you are looking to use, and then find the UUID (as well as the specs) associated with that custom Eve Characteristic.

**Reply by danricho** [reply]
Yes that makes sense. Thanks!

It's a shame there isn't a simple method to creat a similar characteristic to, in this example, brightness, which won't clash with the built in version.

If I use a random long form (Non-Apple) UUID do you think it will fix the non-responsive issue?  eg:

```cpp
// in 'Characteristics.h'

// added:
HAPCHAR( BrightnessSaved, "bd3554cb-fc6e-4554-81d4-5232b7476795", PR+PW+EV, INT, false );
// to line 74 
```
I'm less concerned with the Eve visibility than enabling the NVS and retaining stability.

**Reply by danricho** [reply]
I tried this and it continued the instability so I've reverted to a variable which is saved to NVS

```cpp
// after Homespan accessories are created:
  size_t len;
  nvs_open("SAVED",NVS_READWRITE,&savedNVS);             // open LIGHTS NVS
  if(!nvs_get_blob(savedNVS,"SAVEDDATA",NULL,&len))       // if data found
    nvs_get_blob(savedNVS,"SAVEDDATA",&ledStrip->saved,&len);   // retrieve data
```
```cpp
// any time the saved/favourite is set:
  nvs_set_blob(savedNVS,"SAVEDDATA",&ledStrip->saved,sizeof(ledStrip->saved));        // update data
  nvs_commit(savedNVS);                                                   // commit to NVS
```
